[clang-tidy] Add C++ member function support to custom bugprone-unsafe-functions matches

Discookie opened the pull request:
Before, C++ member functions in the format of ``Class instance; instance.memberfn();`` were unable to be matched.
This PR adds support for this type of call, and it is matched in exactly the same format as other functions (eg. ``::Class::memberfn`` qualified name).

Discookie committed:
[clang-tidy] Add C++ member function support to user-defined bugprone-unsafe-functions matches

PiotrZSL commented on clang-tools-extra/test/clang-tidy/checkers/bugprone/unsafe-functions-custom-regex.cpp:
```diff
@@ -42,3 +47,19 @@ void f1() {
+
+void f2() {
+  S s;
+
```
add test with free standing function member_match_1 that should be allowed.

Discookie replied:
Added.

PiotrZSL reviewed (changes requested):
This isn't good, because "::open" and "std::fstream::open" are 2 different things.
Please make sure that those 2 can be distinguished (configured to detect only one or other or both).

Discookie committed:
Add option to show fully qualified names of functions

5chmidti commented on clang-tools-extra/clang-tidy/bugprone/UnsafeFunctionsCheck.h:
```diff
@@ -43,7 +43,10 @@ class UnsafeFunctionsCheck : public ClangTidyCheck {
   const std::vector<CheckedFunction> CustomFunctions;
 
-  // If true, the default set of functions are reported.
+  /// If true, the fully qualified name of custom functions will be shown in a
+  /// note tag.
+  const bool ShowFullyQualifiedNames;
```
I don't think this is something that should be added. The only reason this exists, is to provide users the capability to debug false-positive matches on their regex. Or is there an additional reason to add this option?

Discookie replied:
Sometimes it is pretty difficult to discern the precise patteren in the first place. eg. above when matching .open(), aliases are not followed (ifstream =/= basic_ifstream<char>), and there's also a template parameter on the class name, but not on the function.

These are existing behaviors, and I haven't found a way to strip template data off of the classname yet.
I'll add some examples for template arguments to the docs.

5chmidti reviewed (commented):
The matching-on-members part looks good, but I don't know about adding the `ShowFullyQualifiedNames` option

Discookie committed:
Add note about matching C++ member functions

HerrCai0907 reviewed (commented):
My first impression is this PR mixes two things. Besides supporting member function, it also add some new flag for it. Could you split it?
For supporting member function, the concept is fine for me.

Discookie committed:
Remove ShowFullyQualifiedNames to be added separately

Discookie committed:
Add example for wildcard template maching in the docs

Discookie commented:
Removed ``ShowFullyQualifiedNames`` for this PR, once this PR is in I'll open a separate one for it.
Added an example for matching all templates to the docs as well.

5chmidti commented on clang-tools-extra/docs/clang-tidy/checks/bugprone/unsafe-functions.rst:
```diff
@@ -139,6 +150,12 @@ Options
     function, and an optional reason, separated by comma. For more information,
     see :ref:`Custom functions<CustomFunctions>`.
 
+.. option:: ShowFullyQualifiedNames
+
+    When `true`, the fully qualified name of all functions matched by the custom
+    function matchers will be shown.
+    Default is `false`.
+
```
This can be removed, now that the option is no longer part of this PR

Discookie committed:
Remove leftover reference to FQ names

Discookie committed:
Merge branch 'main' into unsafe-member-functions
